posts/zh-Hans/2023-06-29-rapid-startup-instanton.adoc

```diff
@@ -0,0 +1,267 @@
+  org.opencontainers.image.version="$VERSION" \
+  org.opencontainers.image.revision="$REVISION" \
+  vendor="Open Liberty" \
+  name="system" \
+  version="$VERSION-$REVISION" \
+  summary="The system microservice from the Getting Started guide" \
+  description="This image contains the system microservice running with the Open Liberty runtime."
+
+COPY --chown=1001:0 src/main/liberty/config/ /config/
+COPY --chown=1001:0 target/*.war /config/apps/
+
+RUN configure.sh
+RUN checkpoint.sh afterAppStart
+----
+
+此配置将应用程序进程检查点添加为应用程序容器镜像的最后一层。`checkpoint.sh`脚本允许指定'afterAppStart'或'beforeAppStart'，以指示在启动的哪个阶段执行进程检查。
+
+两个选项可选来确定检查点发生在应用程序启动之前还是之后：
+
+- `beforeAppStart`: 检查点发生在处理配置的应用程序元数据之后。如果应用程序有作为应用程序启动的一部分运行的任何组件，则在执行应用程序中的任一代码之前采取检查点。这个选项是InstantOn提供的最早检查点阶段。
+- `afterAppStart`: 此选项是检查点能够发生的最后阶段，因此在恢复应用程序实例时，它有可能提供最快的启动时间。检查点发生在所有已配置的应用程序报告为已启动之后。它发生在打开任何监听端口以侦听应用程序的传入请求之前。
+
+`afterAppStart`阶段会为应用程序提供最快的启动时间，但它也可能导致一些应用程序代码在服务器进程检查点发生之前运行。由于本教程中使用的`getting-started`应用程序在其启动逻辑中没有做任何会导致恢复问题的事情，因此我们可以选择使用`afterAppStart` 。
+
+为了InstantOn创建检查点并恢复进程，CRIU二进制文件需要额外的link:/docs/latest/instanton.html#linux-capabilities[Linux功能]。虽然Open Liberty容器镜像包含赋予二进制文件的必要功能。但是，容器在启动时还必须要赋这些功能。
+
+使用podman，您可以使用`-–cap-add`和`--security-opt`选项为容器构建赋予必要的功能，以便在容器构建步骤中创建检查点。启动Podman容器的用户必须具有赋予它必要的Linux功能的权限，因此必须以root或sudo身份运行以下命令：
+
+[source,console]
+----
+podman build \
+   -t dev.local/getting-started-instanton \
+   --cap-add=CHECKPOINT_RESTORE \
+   --cap-add=SYS_PTRACE\
+   --cap-add=SETPCAP \
+   --security-opt seccomp=unconfined .
+----
+
+Dockerfile中的最后一条指令是运行`checkpoint.sh`脚本。当您执行前面的Podman构建命令时，它会启动Open Liberty并在Dockerfile中指定的阶段执行检查点。容器进程数据持久化后，Open Liberty停止，容器镜像构建完成。生成的应用程序容器镜像包含检查点进程数据作为容器镜像的最后一层。输出如下面的例子:
+
+[source,console]
+----
+Performing checkpoint --at=afterAppStart
+
+Launching defaultServer (Open Liberty 23.0.0.6/wlp-1.0.78.cl230620230612-1100) on Eclipse OpenJ9 VM, version 11.0.19+7 (en_US)
+[AUDIT   ] CWWKE0001I: The server defaultServer has been launched.
+[AUDIT   ] CWWKG0093A: Processing configuration drop-ins resource: /opt/ol/wlp/usr/servers/defaultServer/configDropins/defaults/keystore.xml
+[AUDIT   ] CWWKG0093A: Processing configuration drop-ins resource: /opt/ol/wlp/usr/servers/defaultServer/configDropins/defaults/open-default-port.xml
+[AUDIT   ] CWWKZ0058I: Monitoring dropins for applications.
+[AUDIT   ] CWWKZ0001I: Application guide-getting-started started in 1.886 seconds.
+[AUDIT   ] CWWKC0451I: A server checkpoint "afterAppStart" was requested. When the checkpoint completes, the server stops.
+----
+
+[#run]
+==== 运行InstantOn应用程序映像
+
+使用以下命令运行`getting-started-instanton`容器:
+
+[source,console]
+----
+podman run \
+  --rm \
+  --cap-add=CHECKPOINT_RESTORE \
+  --cap-add=SETPCAP \
+  --security-opt seccomp=unconfined \
+  -p 9080:9080 \
+  getting-started-instanton
+----
+
+`--cap-add`的选项赋予容器两个Linux功能，这两个功能是CRIU恢复应用程序进程所需的。当Open Liberty恢复应用程序进程时，它会输出以下消息:
```

Comment:
here also

Reply:
See https://github.com/OpenLiberty/blogs/pull/3909

Reply:
Looks like the JP blog post has an issue with this particular one too.  Unsure my PR will fix everything but we will see.  May be it's all because of the link.